### PR TITLE
Fix #1501 port context/User to nextJS

### LINKED ISSUE #1501: [next] Port contexts/User/ to NextJS
**What would you like to be added**:
Port all components in User/ to NextJS using TS
**Why would you like this to be added**:
Porting reason

### PR DESCRIPTION
**Pull request by humphd**
## Issue This PR Addresses

Fixes #1501

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This picks up the work begun in #1521 and completes it.  I've reworked our User Context in a few ways:

- I've dropped the reducer and user state, preferring to use the `useSWR` hook to manage the network requests.
- I've added a new `User` type, which also includes a boolean property, `isLoggedIn`, which all users have, whether authenticated or not
- I've wrapped our app (i.e., `_app.tsx`) in the User Context, and created a new `useuser()` hook, that any code can use to access the current user.  For example:
```ts
import { useUser } from '../components/UserProvider';

const Component = () => {
  const user = useUser();

  return user.isLoggedIn ? <p>Hi {user.name}</p> : <p>Hi</p>;
};
```

I haven't fully tested this for the authenticated case yet.  I'm going to do that once the login buttons are in.

### CODE REVIEW

**Comment by PedroFonsecaDEV** [comment]
@humphd 
LGTM, I believe everything is working. 
But should we wait until you fully test it before merge?

**Review by PedroFonsecaDEV — APPROVED** [review]

**Review by manekenpix — APPROVED** [review]

**Pull request merged**
